### ISSUE TITLE
Put feature map visualizations in the same "exp" directory

### ISSUE THREAD

**Issue by adrianholovaty** [ultralytics/yolov5#3902]
I was excited to check out the new [feature map functionality](https://github.com/ultralytics/yolov5/pull/3804)! One thing came up in my testing, which seems like an opportunity for improvement.

At the moment, each distinct feature map image gets put in a new (incremented) `exp` directory — `exp2`, `exp3`, `exp4`, etc. It looks as though the feature map image files will always have distinct names within the context of a single run, so it would be safe (and more intuitive) to put them in a single directory together.

So, instead of this resulting directory structure:

```
exp/
    stage_1_Conv_features.png
exp2/
    stage_2_C3_features.png
exp3/
    stage_3_Conv_features.png
exp4/
    stage_4_C3_features.png
exp5/
    stage_5_Conv_features.png
exp6/
    stage_6_C3_features.png
exp7/
    stage_13_C3_features.png
exp8/
    stage_14_Conv_features.png
exp9/
    stage_17_C3_features.png
exp10/
    stage_18_Conv_features.png
exp11/
    stage_20_C3_features.png
```

...it would be:

```
exp/
    stage_1_Conv_features.png
    stage_2_C3_features.png
    stage_3_Conv_features.png
    stage_4_C3_features.png
    stage_5_Conv_features.png
    stage_6_C3_features.png
    stage_13_C3_features.png
    stage_14_Conv_features.png
    stage_17_C3_features.png
    stage_18_Conv_features.png
    stage_20_C3_features.png
```

The downside of the current approach is: When you run inference, the system will generate several `exp` directories in the`features` directory — e.g., `features/exp2`, `features/exp3`, `features/exp4` — while only using a single `detect/exp` directory for the inference results. This makes it difficult to cross-reference which `features/*` directories go with which `detect/*` directories, because the numbers are out of sync.

To see what I mean, change `forward_once()` in `models/yolo.py` to something like this:

```!python
  if feature_vis and m.type in {'models.common.C3', 'models.common.SPP', 'models.common.Conv'}:
      feature_visualization(x, m.type, m.i)
```

I believe this is what @Zigars was referring to in [this comment](https://github.com/ultralytics/yolov5/pull/3807#issuecomment-869657316), but I'm not 100% sure.

The root of this issue is that the new `feature_visualization()` function *always* increments a new `exp` directory, for every image it creates. A solution could be:

* Change `feature_visualization()` to take an `output_dir` parameter and use that, instead of calling `increment_path()`.
* Change all calls to `feature_visualization()` to pass the `output_dir`.
* From there, I'm unsure of the best approach. It seems reasonable to create a global "run number" at the start of training/inference, and pass that to the various places appropriately (the `X` to be used in `detect/expX` and `features/expX`), but I'm not comfortable enough with the YOLOv5 code to know where to put this.

**Comment by glenn-jocher** [comment]
@adrianholovaty ya maybe that's a good idea, though the model of course has no concept of what it's doing really beyond being in .train() mode or .eval() mode, so it doesn't know if it's being called by train, test, detect, or as a hub model, or even less what directories those processes may log to.

I'll go take a look.

**Comment by adrianholovaty** [comment]
@glenn-jocher Yeah, it smells like some tight coupling for sure. :-/

Is there an object that contains state for the current train/test/detect run? Like, runtime params. That would be a good place to put this data, rather than on the model (which, yeah, seems inappropriate).

**Comment by glenn-jocher** [comment]
@adrianholovaty can you try this variant of `feature_visualization()` (in utils/plots.py) and submit a PR if this works better?

```python
def feature_visualization(x, module_type, stage, n=64):
    """
    x:              Features to be visualized
    module_type:    Module type
    stage:          Module stage within model
    n:              Maximum number of feature maps to plot
    """
    batch, channels, height, width = x.shape  # batch, channels, height, width
    if height > 1 and width > 1:
        f = f"stage_{stage}_{module_type.split('.')[-1]}_features.png"  # filename
        project, name = Path('runs/features'), 'exp'
        save_dir = increment_path(project / name, exist_ok=not (project / name / f).exists())  # increment run
        save_dir.mkdir(parents=True, exist_ok=True)  # make dir

        plt.figure(tight_layout=True)
        blocks = torch.chunk(x, channels, dim=1)  # block by channel dimension
        n = min(n, len(blocks))
        for i in range(n):
            feature = transforms.ToPILImage()(blocks[i].squeeze())
            ax = plt.subplot(int(math.sqrt(n)), int(math.sqrt(n)), i + 1)
            ax.axis('off')
            plt.imshow(feature)  # cmap='gray'

        print(f'Saving {save_dir / f}...')
        plt.savefig(save_dir / f, dpi=300)
```

**Comment by glenn-jocher** [comment]
@adrianholovaty good news 😃! Your original issue may now be fixed ✅ in PR #3920. This is a complete revamp of feature visualization with many fixes and improvements.

![stage0_Focus_features](https://user-images.githubusercontent.com/26833433/124769885-25310980-df3a-11eb-995e-3d6fc5f9a4a6.png)

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by adrianholovaty** [comment]
@glenn-jocher Looks good — that fixes the issue I raised here. Nice one!

**Comment by chelsea456** [comment]
> @adrianholovaty good news 😃! Your original issue may now be fixed ✅ in PR #3920. This is a complete revamp of feature visualization with many fixes and improvements.
> 
> ![stage0_Focus_features](https://user-images.githubusercontent.com/26833433/124769885-25310980-df3a-11eb-995e-3d6fc5f9a4a6.png)
> 
> To receive this update:
> 
> * **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
> * **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
> * **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5)
> * **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image [![Docker Pulls](https://camo.githubusercontent.com/280faedaf431e4c0c24fdb30ec00a66d627404e5c4c498210d3f014dd58c2c7e/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f796f6c6f76353f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/yolov5)
> 
> Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!



> @adrianholovaty good news 😃! Your original issue may now be fixed ✅ in PR #3920. This is a complete revamp of feature visualization with many fixes and improvements.
> 
> ![stage0_Focus_features](https://user-images.githubusercontent.com/26833433/124769885-25310980-df3a-11eb-995e-3d6fc5f9a4a6.png)
> 
> To receive this update:
> 
> * **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
> * **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
> * **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5)
> * **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image [![Docker Pulls](https://camo.githubusercontent.com/280faedaf431e4c0c24fdb30ec00a66d627404e5c4c498210d3f014dd58c2c7e/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f796f6c6f76353f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/yolov5)
> 
> Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

Hi sir, i have a question in your picture. In picture hava 32 images, what mean in it's image ? tks sir !!!
